File: apps/backend/src/revolt_api/geo.py
"""Geodesy and route geometry shared by Phase 2 ENC validation.

Deliberate parity with apps/frontend/src/lib/geo.ts: the server-side hazard check needs to reason
about the same path the map draws (straight legs joined by turn-radius arcs), or the "authoritative"
check could disagree with what the operator sees on screen. Ported formula-for-formula rather than
sharing code across languages -- keep this in sync with geo.ts if either changes.
"""

import math
from collections.abc import Sequence
from dataclasses import dataclass

EARTH_RADIUS_M = 6_371_000.0

# Turns tighter than this are treated as effectively straight -- mirrors geo.ts's MIN_TURN_DEG.
_MIN_TURN_DEG = 2.0
# Points sampled along each arc -- mirrors geo.ts's ARC_SEGMENTS.
_ARC_SEGMENTS = 16


def haversine_distance_m(a_lat: float, a_lon: float, b_lat: float, b_lon: float) -> float:
	d_lat = math.radians(b_lat - a_lat)
	d_lon = math.radians(b_lon - a_lon)
	lat1 = math.radians(a_lat)
	lat2 = math.radians(b_lat)
	h = math.sin(d_lat / 2) ** 2 + math.cos(lat1) * math.cos(lat2) * math.sin(d_lon / 2) ** 2
	return 2 * EARTH_RADIUS_M * math.asin(math.sqrt(h))


def bearing_deg(a_lat: float, a_lon: float, b_lat: float, b_lon: float) -> float:
	lat1 = math.radians(a_lat)
	lat2 = math.radians(b_lat)
	d_lon = math.radians(b_lon - a_lon)
	y = math.sin(d_lon) * math.cos(lat2)
	x = math.cos(lat1) * math.sin(lat2) - math.sin(lat1) * math.cos(lat2) * math.cos(d_lon)
	return (math.degrees(math.atan2(y, x)) + 360) % 360


def _normalize_deg_180(deg: float) -> float:
	return ((deg + 180) % 360 + 360) % 360 - 180


def destination_point(
	lat: float, lon: float, bearing_degrees: float, distance_m: float
) -> tuple[float, float]:
	"""Great-circle destination point given a start position, bearing, and distance."""
	delta = distance_m / EARTH_RADIUS_M
	theta = math.radians(bearing_degrees)
	phi1 = math.radians(lat)
	lambda1 = math.radians(lon)
	phi2 = math.asin(
		math.sin(phi1) * math.cos(delta) + math.cos(phi1) * math.sin(delta) * math.cos(theta)
	)
	lambda2 = lambda1 + math.atan2(
		math.sin(theta) * math.sin(delta) * math.cos(phi1),
		math.cos(delta) - math.sin(phi1) * math.sin(phi2),
	)
	return math.degrees(phi2), _normalize_deg_180(math.degrees(lambda2))


def compute_turn_arc(
	prev: tuple[float, float],
	turn: tuple[float, float],
	next_: tuple[float, float],
	radius_m: float,
) -> list[tuple[float, float]] | None:
	"""The circular-arc fillet a vessel follows through a waypoint, tangent to both the inbound and
	outbound legs. Returns None when the turn is negligible or the radius is non-positive -- the
	straight-line corner is already an accurate picture. Mirrors geo.ts's computeTurnArc exactly."""
	if radius_m <= 0:
		return None

	prev_lat, prev_lon = prev
	turn_lat, turn_lon = turn
	next_lat, next_lon = next_

	bearing_in = bearing_deg(prev_lat, prev_lon, turn_lat, turn_lon)
	bearing_out = bearing_deg(turn_lat, turn_lon, next_lat, next_lon)
	turn_delta = _normalize_deg_180(bearing_out - bearing_in)
	if abs(turn_delta) < _MIN_TURN_DEG:
		return None

	leg_in_m = haversine_distance_m(prev_lat, prev_lon, turn_lat, turn_lon)
	leg_out_m = haversine_distance_m(turn_lat, turn_lon, next_lat, next_lon)
	tangent_m = radius_m * math.tan(math.radians(abs(turn_delta)) / 2)
	max_tangent_m = 0.45 * min(leg_in_m, leg_out_m)
	if tangent_m > max_tangent_m:
		tangent_m = max_tangent_m
	if tangent_m <= 0:
		return None

	tangent_in = destination_point(turn_lat, turn_lon, bearing_in + 180, tangent_m)
	turn_sign = 1 if turn_delta >= 0 else -1
	center_bearing = bearing_in + turn_sign * 90
	center = destination_point(tangent_in[0], tangent_in[1], center_bearing, radius_m)
	bearing_center_to_tangent_in = center_bearing + 180

	points: list[tuple[float, float]] = []
	for i in range(_ARC_SEGMENTS + 1):
		sweep = (turn_delta * i) / _ARC_SEGMENTS
		points.append(
			destination_point(center[0], center[1], bearing_center_to_tangent_in + sweep, radius_m)
		)
	return points


@dataclass(frozen=True)
class RouteWaypoint:
	lat: float
	lon: float
	switch_radius: float


def build_route_points(waypoints: Sequence[RouteWaypoint]) -> list[tuple[float, float]]:
	"""The actual path a vessel follows: straight legs joined by turn-radius arcs at each interior
	waypoint (cutting the corner), not the naive straight-line-through-every-waypoint polyline.
	Used to build the LineString that Phase 2 buffers and checks against charted hazards, so the
	server-side check reasons about the same path the map draws rather than a cruder approximation
	of it."""
	if len(waypoints) < 2:
		return [(w.lat, w.lon) for w in waypoints]

	points: list[tuple[float, float]] = [(waypoints[0].lat, waypoints[0].lon)]
	for i in range(1, len(waypoints) - 1):
		prev, turn, nxt = waypoints[i - 1], waypoints[i], waypoints[i + 1]
		arc = compute_turn_arc(
			(prev.lat, prev.lon), (turn.lat, turn.lon), (nxt.lat, nxt.lon), turn.switch_radius
		)
		if arc is None:
			points.append((turn.lat, turn.lon))
		else:
			points.extend(arc)
	points.append((waypoints[-1].lat, waypoints[-1].lon))
	return points

File: apps/frontend/src/lib/geo.ts
import type { Waypoint } from "@revolt/shared-types";

const EARTH_RADIUS_M = 6_371_000;
export const METERS_PER_NM = 1852;
export const METERS_PER_SECOND_TO_KNOTS = 1.94384;

function toRad(deg: number): number {
	return (deg * Math.PI) / 180;
}

function toDeg(rad: number): number {
	return (rad * 180) / Math.PI;
}

/** Great-circle distance between two lat/lon points, in meters (haversine formula). */
export function haversineDistanceM(aLat: number, aLon: number, bLat: number, bLon: number): number {
	const dLat = toRad(bLat - aLat);
	const dLon = toRad(bLon - aLon);
	const lat1 = toRad(aLat);
	const lat2 = toRad(bLat);
	const h = Math.sin(dLat / 2) ** 2 + Math.cos(lat1) * Math.cos(lat2) * Math.sin(dLon / 2) ** 2;
	return 2 * EARTH_RADIUS_M * Math.asin(Math.sqrt(h));
}

/** Initial great-circle bearing from point a to point b, in degrees, normalized to [0, 360). */
export function bearingDeg(aLat: number, aLon: number, bLat: number, bLon: number): number {
	const lat1 = toRad(aLat);
	const lat2 = toRad(bLat);
	const dLon = toRad(bLon - aLon);
	const y = Math.sin(dLon) * Math.cos(lat2);
	const x = Math.cos(lat1) * Math.sin(lat2) - Math.sin(lat1) * Math.cos(lat2) * Math.cos(dLon);
	return (toDeg(Math.atan2(y, x)) + 360) % 360;
}

function normalizeDeg180(deg: number): number {
	return ((((deg + 180) % 360) + 360) % 360) - 180;
}

/**
 * Great-circle destination point given a start position, bearing, and distance: the standard
 * "direct geodesic" formula (same family as haversineDistanceM/bearingDeg above).
 */
export function destinationPoint(
	lat: number,
	lon: number,
	bearingDegrees: number,
	distanceM: number,
): { lat: number; lon: number } {
	const delta = distanceM / EARTH_RADIUS_M;
	const theta = toRad(bearingDegrees);
	const phi1 = toRad(lat);
	const lambda1 = toRad(lon);
	const phi2 = Math.asin(
		Math.sin(phi1) * Math.cos(delta) + Math.cos(phi1) * Math.sin(delta) * Math.cos(theta),
	);
	const lambda2 =
		lambda1 +
		Math.atan2(
			Math.sin(theta) * Math.sin(delta) * Math.cos(phi1),
			Math.cos(delta) - Math.sin(phi1) * Math.sin(phi2),
		);
	return { lat: toDeg(phi2), lon: normalizeDeg180(toDeg(lambda2)) };
}

/** One end of a route leg. */
export interface LegEndpoint {
	id: string;
	lat: number;
	lon: number;
}

/** One leg of the route, from one waypoint to the next. */
export interface LegPositions {
	toId: string;
	from: LegEndpoint;
	to: LegEndpoint;
}

/** Substitutes one waypoint's live position (e.g. mid-drag) ahead of it being persisted. */
export interface PositionOverride {
	id: string;
	lat: number;
	lon: number;
}

/**
 * Consecutive from/to endpoint pairs for each leg of the route, in sequence order.
 * `override` substitutes one waypoint's live position (e.g. mid-drag, before it's
 * persisted) so callers can recompute geometry/hazards without touching context state.
 */
export function computeLegPositions(
	waypoints: Waypoint[],
	override?: PositionOverride,
): LegPositions[] {
	const endpoint = (w: Waypoint): LegEndpoint =>
		override?.id === w.id
			? { id: w.id, lat: override.lat, lon: override.lon }
			: { id: w.id, lat: w.position.latitude, lon: w.position.longitude };

	const legs: LegPositions[] = [];
	for (let i = 1; i < waypoints.length; i++) {
		const from = waypoints[i - 1];
		const to = waypoints[i];
		if (!from || !to) continue;
		legs.push({ toId: to.id, from: endpoint(from), to: endpoint(to) });
	}
	return legs;
}

// Turns tighter than this are treated as effectively straight: not worth drawing an arc for,
// and the tangent-length formula below blows up as the turn approaches a full reversal anyway.
const MIN_TURN_DEG = 2;
// Points sampled along each arc; enough to look smoothly curved at chart zoom levels without
// generating an excessive number of GeoJSON vertices per waypoint.
const ARC_SEGMENTS = 16;

/** One waypoint's turn-radius fillet, ready to render as a map line. */
export interface TurnArc {
	waypointId: string;
	points: { lat: number; lon: number }[];
}

/**
 * The circular-arc "fillet" a vessel actually follows through a waypoint, tangent to both the
 * inbound and outbound legs: the same construction real ECDIS route planning uses to turn a
 * per-waypoint turn radius into a course-change arc (paired with a wheel-over point marking where
 * the turn begins, not modeled here). Returns null when the turn is negligible or the radius is
 * zero; nothing to draw, the straight-line corner is already an accurate picture.
 */
export function computeTurnArc(
	prev: { lat: number; lon: number },
	turn: { lat: number; lon: number },
	next: { lat: number; lon: number },
	radiusM: number,
): { lat: number; lon: number }[] | null {
	if (radiusM <= 0) return null;

	const bearingIn = bearingDeg(prev.lat, prev.lon, turn.lat, turn.lon);
	const bearingOut = bearingDeg(turn.lat, turn.lon, next.lat, next.lon);
	const turnDelta = normalizeDeg180(bearingOut - bearingIn);
	if (Math.abs(turnDelta) < MIN_TURN_DEG) return null;

	const legInM = haversineDistanceM(prev.lat, prev.lon, turn.lat, turn.lon);
	const legOutM = haversineDistanceM(turn.lat, turn.lon, next.lat, next.lon);
	// Tangent length from the waypoint to where the arc leaves each straight leg: the standard
	// circular-fillet formula (radius * tan(halfAngle)), the same math behind wheel-over point
	// calculations.
	let tangentM = radiusM * Math.tan(toRad(Math.abs(turnDelta)) / 2);
	// Never eat more than 45% of either adjacent leg, so arcs at closely-spaced waypoints don't
	// overshoot past the next/previous waypoint or overlap a neighboring arc.
	const maxTangentM = 0.45 * Math.min(legInM, legOutM);
	if (tangentM > maxTangentM) tangentM = maxTangentM;
	if (tangentM <= 0) return null;

	const tangentIn = destinationPoint(turn.lat, turn.lon, bearingIn + 180, tangentM);
	// Center of the fillet circle sits perpendicular to the inbound leg at tangentIn, on the side
	// the turn curves toward.
	const turnSign = turnDelta >= 0 ? 1 : -1;
	const centerBearing = bearingIn + turnSign * 90;
	const center = destinationPoint(tangentIn.lat, tangentIn.lon, centerBearing, radiusM);
	const bearingCenterToTangentIn = centerBearing + 180;

	const points: { lat: number; lon: number }[] = [];
	for (let i = 0; i <= ARC_SEGMENTS; i++) {
		const sweep = (turnDelta * i) / ARC_SEGMENTS;
		points.push(
			destinationPoint(center.lat, center.lon, bearingCenterToTangentIn + sweep, radiusM),
		);
	}
	return points;
}

/**
 * Turn arcs for every interior waypoint (one with both a previous and next neighbor) that has a
 * non-negligible course change and a positive switch_radius. Endpoints of the route never get an
 * arc: there's no incoming or outgoing leg to blend.
 */
export function computeTurnArcs(waypoints: Waypoint[], override?: PositionOverride): TurnArc[] {
	const position = (w: Waypoint): { lat: number; lon: number } =>
		override?.id === w.id
			? { lat: override.lat, lon: override.lon }
			: { lat: w.position.latitude, lon: w.position.longitude };

	const arcs: TurnArc[] = [];
	for (let i = 1; i < waypoints.length - 1; i++) {
		const prev = waypoints[i - 1];
		const turn = waypoints[i];
		const next = waypoints[i + 1];
		if (!prev || !turn || !next) continue;
		const points = computeTurnArc(
			position(prev),
			position(turn),
			position(next),
			turn.switch_radius,
		);
		if (points) arcs.push({ waypointId: turn.id, points });
	}
	return arcs;
}
